### PR TITLE
[bug 1169261] Implement redirector

### PR DESCRIPTION
**Pull request by willkg**
* implement redirector app
* add google_utils for pinging the server with events
* wrote SUMO Suggest redirector
* rework SUMO Suggester to use the redirector
* deal with the AAQ link
* minor renaming of things because I like to rename things over and over
  again when it's raining

r?

### CODE REVIEW

**Comment by willkg** [comment]
The redirector is an app that manages the `/redirect` view. It takes an `r` parameter which it then passes through a list of redirector providers as specified in `settings.REDIRECTOR_PROVIDERS`. The first one to return a url wins and then the user is redirected to that url.

This creates a SUMO Suggest redirector. This encodes the rank of the url to redirect to or "aaq" in the redirection value. The rank is the index of the document stored in the user's session which was returned from the SUMO Suggest API.

In this way:

1. no one can see anyone else's redirections
2. redirection can't (currently) be abused to send users to "naughty places"
3. redirections and documents expire with the django session
4. we capture the appropriate information and send it to google so we can see event flows and compare "# suggests" to "# number of aaq links pressed" and "# non-aaq links pressed" and use this to tune the suggestion algorithm and/or even know whether it's useful at all

That's the gist of it.

File: fjord/redirector/base.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        """Implement this to handle redirects for a given request
+
+        :returns: None or url to redirect to
+
+        """
+        raise NotImplementedError
+
+
+def build_redirect_url(redirect):
+    return reverse('redirect-view') + '?' + urllib.urlencode({'r': redirect})
```
Comment (rlr) [inline fjord/redirector/base.py]: should this use `fjord.base.helpers.urlparams()` instead?

File: fjord/suggest/providers/sumosuggest.py [hunk 2]
```diff
@@ -46,11 +54,115 @@ def get_kb_articles(locale, product, text):
...
+        # Extract the rank.
+        try:
+            trigger, rank = parse_redirect(redirect)
+        except RedirectParseError:
+            return
+
+        if rank == 'aaq':
+            url = SUMO_AAQ_URL
+        else:
+            # This doc doesn't exist.
```
Comment (rlr) [inline fjord/suggest/providers/sumosuggest.py]: this comment is confusing. does it go inside the `except` below?

**Comment by rlr** [comment]
I wasn't expecting the session to be used, that's awesome for all the reasons you noted. I'm going to chew on that for a little but I think I am :+1: to that.

File: fjord/redirector/base.py (hunk 1 above)
Comment (willkg) [inline fjord/redirector/base.py]: Mmm... `urlparams` is a template filter and it does a bunch of other stuff like handling duplicates that we don't need. I think I'm on the fence.

File: fjord/suggest/providers/sumosuggest.py (hunk 2 above)
Comment (willkg) [inline fjord/suggest/providers/sumosuggest.py]: Whoops--you're right.

**Comment by willkg** [comment]
The other ways to do it:

1. store suggestions in the db which requires us to do some more implementation and also write some code to clean them up later
2. store suggestions in cache, but i'm concerned that it'll get bumped out of cache which would cause a bunch of problems with the redirects

Are there other options? Seems like for something like this where it's a short list of dicts, putting it in the session works.

**Comment by willkg** [comment]
I was waiting for the microwave to reheat my frittata and mulling over everything when it occurred to me that if you leave two feedbacks with the same django session, then the sumo suggest provider will show you suggestions from the first feedback because it's not storing the response id. Pretty sure that's true. I'll look into it now.

Also interesting, the spelling-correct thing wants to replace "frittata" with "fritter".

**Comment by willkg** [comment]
^^^ That fixes the "comment in the wrong place" problem plus my "sumosuggest provider should be cleverer about docs in the session" problem.

**Comment by rlr** [comment]
I like. Gonna run tests since travis doesn't like us today

File: fjord/suggest/tests/test_sumosuggest.py [hunk 3]
```diff
@@ -97,86 +157,113 @@ def test_suggestions(self):
         }
 
         with requests_mock.Mocker() as m:
-            m.get(SUMO_SUGGEST_API_URL, json=ret)
+            m.get(SUMO_SUGGEST_API_URL, json=sumo_api_ret)
+            m.post(GOOGLE_API_URL, text='whatever')
```
Comment (rlr) [inline fjord/suggest/tests/test_sumosuggest.py]: I'm digging the requests mock. Very cool

**Comment by rlr** [comment]
I ran tests and read the code again. r+

**Comment by willkg** [comment]
Awesome--thank you!

**Comment by willkg** [comment]
^^^ Rebased.

**Pull request merged**